### PR TITLE
Feat/text features : first implems of text-annotation types to Pixano

### PR DESCRIPTION
**Pull request by benlabbe**
## Description

Add text-annotation types to Pixano for *NamedEntity* , *Relation* and *Classification*.
The later one is agnostic to the View. It can be used for text, image or even video datasets

### CODE REVIEW

**Comment by benlabbe** [comment]
@juliendenize  , @BertrandRenault  , @jrabary  , could we have two of you that review/approve this pull request , so that we could move toward the merge .   Thanks

File: pyproject.toml [hunk 1]
```diff
@@ -25,6 +25,7 @@ classifiers = [
   "License :: CeCILL-C Free Software License Agreement (CECILL-C)",
 ]
 dependencies = [
+  "polars >= 1.9.0, < 2.0.0",
```
Comment (BertrandRenault) [inline pyproject.toml:28]: there is already a polars dependencies line 41

**Review by BertrandRenault — APPROVED** [review]

File: pixano/features/schemas/annotations/classification.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+    Attributes:
+        labels: List of class names.
+        confidences: List of prediction confidences.
+    """
+
+    labels: list[str]
+    confidences: list[float]
+
+    @model_validator(mode="after")
```
Comment (juliendenize) [inline pixano/features/schemas/annotations/classification.py:28]: Could you add unit tests ?

File: pixano/features/schemas/annotations/classification.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            "None" Classification.
+        """
+        return cls(
+            id="",
+            item=ItemRef.none(),
+            view=ViewRef.none(),
+            entity=EntityRef.none(),
+            labels=[],
+            confidences=[],
+        )
```
Comment (juliendenize) [inline pixano/features/schemas/annotations/classification.py:49]: Could you add unit tests ?

File: pixano/features/schemas/annotations/classification.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+    @property
+    def predictions(self) -> list[tuple[str, float]]:
+        """Get list of zipped predictions (labels and confidences)."""
+        return list(zip(self.labels, self.confidences))
```
Comment (juliendenize) [inline pixano/features/schemas/annotations/classification.py:54]: Could you add unit tests ?

File: pixano/features/schemas/annotations/classification.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    """
+    return Classification(
+        labels=labels,
+        confidences=confidences,
+        id=id,
+        item_ref=item_ref,
+        view_ref=view_ref,
+        entity_ref=entity_ref,
+        source_ref=source_ref,
+    )
```
Comment (juliendenize) [inline pixano/features/schemas/annotations/classification.py:93]: Could you add unit tests ?

File: pixano/features/schemas/annotations/info_extraction.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    def _validate_fields(self):
+        if len(self.spans_start) != len(self.spans_end):
+            raise ValueError("Spans offset lists should have the same length")
+        if len(self.spans_start) == 0:
+            warnings.warn("To ground a NamedEntity in a text, spans offsets should not be empty", category=UserWarning)
+        if not all(span[0] >= 0 and span[1] >= 0 for span in zip(self.spans_start, self.spans_end)):
+            raise ValueError("Spans offset must be positive")
+        if not all(span[0] <= span[1] for span in zip(self.spans_start, self.spans_end)):
+            raise ValueError("End offsets must be greater or equal to their corresponding start offset")
+        return self
```
Comment (juliendenize) [inline pixano/features/schemas/annotations/info_extraction.py:45]: Could you add unit tests ?

File: pixano/features/schemas/annotations/info_extraction.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    """Check if a class is a NamedEntity or subclass of NamedEntity."""
+    return issubclass_strict(cls, NamedEntity, strict)
+
+
+def is_relation(cls: type, strict: bool = False) -> bool:
+    """Check if a class is a Relation or subclass of Relation."""
+    return issubclass_strict(cls, Relation, strict)
+
+
+def create_namedentity(
```
Comment (juliendenize) [inline pixano/features/schemas/annotations/info_extraction.py:120]: Rename namedentity to named_entity to keep consistency

File: pixano/features/schemas/annotations/info_extraction.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+            item=ItemRef.none(),
+            view=ViewRef.none(),
+            entity=EntityRef.none(),
+            predicate_id="",
+            subject_id=AnnotationRef.none(),
+            object_id=AnnotationRef.none(),
+        )
+
+
+def is_namedentity(cls: type, strict: bool = False) -> bool:
```
Comment (juliendenize) [inline pixano/features/schemas/annotations/info_extraction.py:110]: Rename namedentity to named_entity to keep consistency

File: pixano/utils/python.py [hunk 9]
```diff
@@ -123,3 +123,10 @@ def fn_sort_dict(dict_: dict[str, Any], order_by: list[str], descending: list[bo
         else:
             key.append(value)
     return tuple(key)
+
+
+def uniq_list(seq: list) -> list:
```
Comment (juliendenize) [inline pixano/utils/python.py:128]: Rename to unique_list

Comment (juliendenize) [inline pixano/utils/python.py:128]: Rename seq to list_ if it only deal with lists or sequence if it deals with sequences

In the second case, change the type to Sequence.

File: tests/features/schemas/annotations/test_information_extraction.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+        assert none_rel.id == ""
+        assert none_rel.item_ref == ItemRef.none()
+        assert none_rel.view_ref == ViewRef.none()
+        assert none_rel.entity_ref == EntityRef.none()
+        assert none_rel.predicate_id == ""
+        assert none_rel.subject_id == AnnotationRef.none()
+        assert none_rel.object_id == AnnotationRef.none()
+
+
+def test_is_namedentity():
```
Comment (juliendenize) [inline tests/features/schemas/annotations/test_information_extraction.py:94]: Rename to test_is_named_entity

File: tests/features/schemas/annotations/test_information_extraction.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+    make_tests_is_sublass_strict(is_namedentity, NamedEntity)
+
+
+def test_is_relation():
+    make_tests_is_sublass_strict(is_relation, Relation)
+
+
+def test_create_namedentiy():
```
Comment (juliendenize) [inline tests/features/schemas/annotations/test_information_extraction.py:102]: rename to test_create_named_entiy

File: pixano/utils/python.py [hunk 12]
```diff
@@ -123,3 +123,10 @@ def fn_sort_dict(dict_: dict[str, Any], order_by: list[str], descending: list[bo
...
+    """Select uniq elements in a list while keeping order."""
+    seen: list = []
+    seen_add = seen.append
+    return [x for x in seq if not (x in seen or seen_add(x))]
```
Comment (juliendenize) [inline pixano/utils/python.py:132]: Could you add unit tests ?

**Review by juliendenize — CHANGES_REQUESTED** [review]

File: pixano/utils/python.py (hunk 12 above)
Comment (benlabbe) [reply pixano/utils/python.py]: test added

File: tests/features/schemas/annotations/test_information_extraction.py (hunk 11 above)
Comment (benlabbe) [reply tests/features/schemas/annotations/test_information_extraction.py]: done

File: pixano/utils/python.py (hunk 9 above)
Comment (benlabbe) [reply pixano/utils/python.py]: done

Comment (benlabbe) [reply pixano/utils/python.py]: done

File: pixano/features/schemas/annotations/info_extraction.py (hunk 6 above)
Comment (benlabbe) [reply pixano/features/schemas/annotations/info_extraction.py]: done

**Review by juliendenize — APPROVED** [review]

**Pull request merged**